### DISCUSSION TITLE
Default, Removable Element Conditions for Element Sources

### DISCUSSION THREAD

**Discussion by myleshyson** [craftcms/cms#13464] [Ideas] [answered]
I have a custom `Event` element with a element index page in the control panel. It's part of a calendar module I built for our healthcare system, and we would like users to be able to, by default, when selecting a calendar source, to see all events scoped to that calendar and filtered by events happening in the current month. 

The calendar scope is currently handled by `defineSources` on an event by setting the calendar in the source criteria. That shouldn't be removable by the user. However, the filtering by the current month _should_ be removable by the user. 

Ideally, there would be a pathway to define default condition rules for an element index, that show up in the FilterHUD when loading the screen, and are removable or changeable if the user wants to filter by something else.

The reason this is a thing for us is because defining a default sort order isn't really helpful. Seeing the oldest events or events happening a year from now isn't ideal. We also show recurrences in that index view, so we really need a good starting point for users to filter by.

I noticed that when defining an element source, you can either define it as `native` or `custom`. When `native`, the element index controller only respects the criteria key. When `custom`, it only respects the `condition` key. What we need is a hybrid, some api that respects both criteria _and_ condition, but that also respects if a user removes that condition.

I currently have a hack way of doing this by defining an `EventIndex.js` module, and more or less tricking the filterHUD into thinking there's already a filter defined. on load. It's definitely one of those things that feels like it will break with any new update from Craft. 

Here's a some images of what I'm talking about. 

On page load or initial source select, events are scoped to the selected calendar, and the filter icon is active to show there's a active filter.
<img width="1389" alt="Screenshot 2023-07-21 at 12 36 36 PM" src="https://github.com/craftcms/cms/assets/11747187/33a2acc3-0ac0-457a-8cc1-c3fd1857e054">

If a opens the FilterHUD, they'll see that we have a default "Start Date" filter that's set to the current month. If they remove that filter, or change it, that source would respect that choice until the next page load. 
<img width="1371" alt="Screenshot 2023-07-21 at 12 38 25 PM" src="https://github.com/craftcms/cms/assets/11747187/a1a718a0-8f46-4666-9d63-471ec949776f">

Any thoughts on this? Not sure exactly yet where the best place to define this. Im thinking maybe updating the element indexes controller to respect "condition" when the type is "native", and updating the BaseElementIndex to respect that default filter on intial page load.

**Comment by myleshyson** [comment]
Also open to feedback if I'm thinking about this all wrong!

**Comment by myleshyson** [comment]
Ok I found a less hacky (but still kinda hacky) way to do this. Still would be nice for there to be support for this. What i did was
1. Make all of the `Event` sources custom, and define both criteria and a condition with a rule.
```php
        $sources[] = [
            'key' => '*',
            'label' => 'All Events',
            'criteria' => [
                'with' => ['calendars'],
                'orderBy' => ['startDate' => SORT_DESC],
            ],
            'condition' =>[
                'class' => EventCondition::class,
                'conditionRules' => [
                    [
                        'class' => StartDateConditionRule::class,
                        'rangeType' => DateRange::TYPE_THIS_MONTH
                    ]
                ]
            ],
            'defaultSort' => ['startDate', 'desc'],
            'type' => 'custom'
        ];
```
2. Added a `source` key to my element query class, which when definied gets the source and just applies the criteria in the source.
```php
    private function _applySource()
    {
        if (!$this->source) {
            return;
        }

        $source = ElementHelper::findSource(Event::class, $this->source);
        if (empty($source['criteria'])) {
            return;
        }

        foreach ($source['criteria'] as $key => $value) {
            $this->$key = $value;
        }
    }
```
3. Create an EventIndex js module and make sure when sending params, it sends the source as a criteria. Also make sure it hides the remove button from the first filter. Decided it doesn't really make sense in this context that the filter should be removed, but it does need to be editable.
```js
/** global: Garnish */
Craft.EventIndex = Craft.BaseElementIndex.extend({
  firstLoad: true,
  observer: null,
  init(elementType, $container, settings) {
    this.base(elementType, $container, settings)
  },
  getViewParams() {
    let params = this.base()
    params.criteria.source = this.sourceKey
    return params
  },
  updateElements() {
    this.base()
    this.$filterBtn.addClass('active')
  },
  /**
   * Craft doesn't fully support default filters just yet. So this does a couple things to make things work properly.
   * 1. If there's no rules (somehow somebody removed the first filter), this resets the filter hud so our default
   *    rule is loaded again.
   * 2. Listens for when any new rules are loaded, and remove the delete button for the first one. We don't want
   *    people removing our default "Start Date" filter. We just want to allow them to update it.
   * @todo make sure the clear button just says "Cancel" if the only rule is our default rule.
   */
  showFilterHud() {
    let $hud = get(this.filterHuds, `${this.siteId}.${this.sourceKey}`, null)
    if ($hud !== null && !$hud.hasRules()) {
      this.filterHuds[this.siteId][this.sourceKey] = null
    }
    this.base()

    if (this.observer) {
      return
    }

    $hud = get(this.filterHuds, `${this.siteId}.${this.sourceKey}`, null)

    this.observer = new MutationObserver((mutationList, observer) => {
      for (const mutation of mutationList) {
        if (!mutation.target.classList.contains('loading')) {
          $(mutation.target.querySelector('#condition-conditionRules-1-rule-actions')).remove()
        }
      }
    })
    this.observer.observe($hud.$hud[0], {
      attributes: true,
      childList: false,
      subtree: false
    })
  }
})

Craft.registerElementIndexClass('modules\\calendar\\elements\\Event', Craft.EventIndex)
```
This works relatively well. Im not a huge fan of relying on the MutationObserver in the js module, or needing a custom way to set source criteria to the element query. Still feels like it might break any moment.

**Comment by brandonkelly** [comment, marked as answer]
You should not be defining `custom` sources programmatically – those are expected to be defined by admins via the Customize Sources modal.

I’ve just added support for native sources to define a `defaultFilter` key, for Craft 4.5, which should help you here (#13499). 

```php
$condition = Event::createCondition();
$condition->setConditionRules([
    new StartDateConditionRule(['rangeType' => DateRange::TYPE_THIS_MONTH]),
]);

$sources[] = [
    'key' => '*',
    'label' => 'All Events',
    'criteria' => [
        'with' => ['calendars'],
        'orderBy' => ['startDate' => SORT_DESC],
    ],
    'defaultFilter' => $condition,
    'defaultSort' => ['startDate', 'desc'],
    'type' => 'custom'
];
```

If you’d like to test this ahead of time, you can change your `craftcms/cms` requirement in `composer.json` to `4.5.x-dev as 4.5.0-alpha`, and then run `composer update`.

**Reply by myleshyson** [reply]
This is PERFECT for our use case. Thank you.

**Reply by brandonkelly** [reply]
Craft 4.5.0 is out with those changes!
